torch/nn/utils/parametrize.py

```diff
@@ -0,0 +1,423 @@
+        original (Parameter or Tensor): parameter or buffer that is parametrized
+    """
+
+    def __init__(
+        self, modules: Iterable[Module], original: Union[Tensor, Parameter]
+    ) -> None:
+        super().__init__(modules)
+        if isinstance(original, Parameter):
+            self.register_parameter("original", original)
+        else:
+            self.register_buffer("original", original)
+
+    def set_original_(self, value: Tensor) -> None:
+        r"""This method is called when assigning to a parametrized tensor.
+        It calls the methods ``right_inverse`` (see :func:`register_parametrization`)
+        of the parametrizations in the inverse order that they have been registered.
+        Then, it assigns the result to ``self.original``.
+
+        Args:
+            value (Tensor): Value to which initialize the module
+        """
+        for module in reversed(self):
+            if not hasattr(module, "right_inverse"):
+                raise RuntimeError(
+                    "The parametrization '{}' does not implement a 'right_inverse' method. "
+                    "Assigning to a parametrized tensor is only possible when all the parametrizations "
+                    "implement a 'right_inverse' method.".format(
+                        module.__class__.__name__
+                    )
+                )
+
+        with torch.no_grad():
+            for module in reversed(self):
+                value = module.right_inverse(value)
+            self.original.copy_(value)
+
+    def forward(self) -> Tensor:
+        x = self.original
+        for module in self:
+            x = module(x)
+        if x.size() != self.original.size():
+            raise RuntimeError(
+                "The parametrization may not change the size of the parametrized tensor. "
+                "Size of original tensor: {} "
+                "Size of parametrized tensor: {}".format(self.original.size(), x.size())
+            )
+        return x
+
+
+def set_caching(module: Module, tensor_name: str) -> None:
+    r"""Sets up the caching mechanism for a given parametrization.
+    After applying this function, the values of the parametrization will be cached
+    when the context manager :func:`~cached` is active.
+    This function is automatically invoked within :func:`register_parametrization`.
+    This function is the inverse of :func:`remove_caching`.
+
+    Args:
+        module (nn.Module): module on which to remove the caching mechanism
+        tensor_name (str): name of the parameter, buffer, or parametrization
+            from which the parametrization will be applied
+    """
+    if not is_parametrized(module, tensor_name):
+        raise ValueError(
+            "The tensor {} in module {} is not parametrized".format(
+                tensor_name, type(module)
+            )
+        )
+
+    key = _key(module, tensor_name)
+    if key not in _cache:
+        _cache[key] = None
```

Comment:
Sure thing. I also think that this is a better design.

Reply:
When writing the code (it's pushed now), I did not know how to go about having the attribute in `ParametrizationList`. If I add it as an attribute, it will not be saved when calling `model.state_dict()`. On the other hand, if it's saved as a boolean buffer, manipulating it is a bit cumbersome as one has to do:
```python
model.parametrizations.weight.is_cached = torch.tensor(False)
```
What do you think? Should we go for the latter one even though it is a bit cumbersome? It is going to be documented, so it should be fine for users regardless.

Reply:
I am not sure to understand why this should be in the state dict?
That would be a property of the model you load things into no? The state dict is only about weights and buffers.

Reply:
Yeah, but I think that this can be very much a property of the model. I cannot think of any example where setting `is_cached=False` is reasonable, but if people want to use it, I would assume that they would like to save it and load it, rather than having to unset it every time they load a model...

Reply:
If we don't know when this feature would be useful, why not just remove it.
That will simplify this PR. And we can add it back when users have actual use case for it.